txpipe/twopoint.py

```diff
@@ -739,6 +740,226 @@ def load_lens_catalog(self, data):
+        ('summary_statistics_fourier', SACCFile),
+        ('fiducial_cosmology', FiducialCosmology),  # For example lines
+    ]
+    outputs = [
+        ('twopoint_theory_real', SACCFile),
+        ('twopoint_theory_fourier', SACCFile),
+        ]
+    
+
+    def run(self):
+        import sacc
+
+        # Real space
+        filename = self.get_input('summary_statistics_real')
+        s = sacc.Sacc.load_fits(filename)
+
+        # TODO: when there is a better Cosmology serialization method
+        # switch to that
+        print("Manually specifying matter_power_spectrum and Neff")
+        cosmo = self.open_input('fiducial_cosmology', wrapper=True).to_ccl(
+        matter_power_spectrum='halofit', Neff=3.04)
+        print(cosmo)
+
+        s_theory, tracers, nbin_source, nbin_lens = self.replace_with_theory_real(s, cosmo)
+        # save the output to Sacc file
+        s_theory.save_fits(self.get_output('twopoint_theory_real'), overwrite=True)
+
+        # Fourier space
+        filename = self.get_input('summary_statistics_fourier')
+        s = sacc.Sacc.load_fits(filename)
+
+        s_theory = self.replace_with_theory_fourier(s, cosmo, tracers, nbin_source, nbin_lens)
+        # save the output to Sacc file
+        s_theory.save_fits(self.get_output('twopoint_theory_fourier'), overwrite=True)
+
+        
+    def read_nbin(self, s):
+        import sacc
+
+        xip = sacc.standard_types.galaxy_shear_xi_plus
+        wtheta = sacc.standard_types.galaxy_density_xi
+
+        source_tracers = set()
+        for b1, b2 in s.get_tracer_combinations(xip):
+            source_tracers.add(b1)
+            source_tracers.add(b2)
+
+        lens_tracers = set()
+        for b1, b2 in s.get_tracer_combinations(wtheta):
+            lens_tracers.add(b1)
+            lens_tracers.add(b2)
+
+
+        return len(source_tracers), len(lens_tracers)
+
+    def get_ccl_tracers(self, s, cosmo, smooth=False):
+        
+        # ccl tracers object
+        import pyccl
+        tracers = {}
+
+        nbin_source, nbin_lens = self.read_nbin(s)
+        
+        # Make the lensing tracers
+        for i in range(nbin_source):
+            name = f'source_{i}'
+            Ti = s.get_tracer(name)
+            nz = smooth_nz(Ti.nz) if smooth else Ti.nz
+            print("smooth:",  smooth)
+            # Convert to CCL form
+            tracers[name] = pyccl.WeakLensingTracer(cosmo, (Ti.z, nz))
+
+        # And the clustering tracers
+        for i in range(nbin_lens):
+            name = f'lens_{i}'
+            Ti = s.get_tracer(name)
+            nz = smooth_nz(Ti.nz) if smooth else Ti.nz
+
+            # Convert to CCL form
+            tracers[name] = pyccl.NumberCountsTracer(cosmo, has_rsd=False, 
+                dndz=(Ti.z, nz), bias=(Ti.z, np.ones_like(Ti.z)))
+            
+        return tracers
+    
+    def replace_with_theory_real(self, s, cosmo):
+
+        import pyccl
+        nbin_source, nbin_lens = self.read_nbin(s)
```

Comment:
is this all copied from the theory.py?  might be better to avoid repeating ourselves and have this in one place. There's a principle D.R.Y. - Don't Repeat Yourself.

Reply:
yeah agreed. I think the best way would be to load the theory files in the plotting stage. Instead of computing the theory there again. Do you agree? I could try to implement a version of that.